### PR TITLE
Enhance stability of Product E2E Tests

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This pull request aims to improve the stability of the product E2E (end-to-end) tests by addressing their occasional flakiness.

Currently, the E2E tests occasionally exhibit instability. The inconsistency primarily arises when testing two distinct methods of creating a product: the legacy approach and the new product experience.

To rectify this situation, this PR introduces changes to execute these tests sequentially instead of in parallel. By doing so, the tests relating to the new product experience and the legacy approach will be conducted at separate instances.

I conducted multiple test runs locally and remotely on this branch, and the outcomes consistently demonstrated that the flakiness was successfully mitigated.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Execute the standard commands to launch the local E2E environment.
2. Make sure the test is not flaky by running multiple instances of it repeatedly. You can do so through this command:
    ```bash
    cd plugins/woocommerce
    pnpm test:e2e-pw ./tests/e2e-pw/tests/merchant/products/ --retries=0 --workers=4 --repeat-each=8
    ```
4. All instances of this test should pass, which indicates it isn't flaky.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/add-variable-product/create-product-attributes.spec.js [hunk 1]
```diff
@@ -23,6 +23,8 @@ const step_goToAttributesTab = async ( page ) => {
 	} );
 };
 
+test.describe.configure( { mode: 'serial' } );
```
Comment (louwie17) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/products/add-variable-product/create-product-attributes.spec.js:26]: @octaedro it seems using this is not recommended per the playwright docs, and it also seems we are the only one's using it within these tests.
Can you add some more details and context on why we need this, and what caused the E2E tests to be flaky?

**Review by louwie17 — COMMENTED** [review]
Left a comment asking about the `serial` mode, also there is not much context on these changes, which makes it hard to determine how this should make the tests more robust or not. 
Could you add some extra context, either in the description or inline on why these changes where necessary? What the exact issue was?
Feel free to link to a Slack thread or something if it was discussed in detail there.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/add-variable-product/create-product-attributes.spec.js (hunk 1 above)
Comment (octaedro) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/add-variable-product/create-product-attributes.spec.js:26]: Thank you @louwie17 for your comments. I'll update the PR's description 🙂


> it seems using this is not recommended per the playwright docs, and it also seems we are the only one's using it within these tests.

Yeah, running tests serially is not recommended. However, in this situation, I think it's necessary. We aim to test two distinct methods of creating a product, and both depend on the feature `product-block-editor`, if we run these tests concurrently, one will likely fail.

Despite we're not using `test.describe.configure`, we use `test.describe.serial` [instead](https://github.com/search?q=repo%3Awoocommerce%2Fwoocommerce+path%3A%2F%5Eplugins%5C%2Fwoocommerce%5C%2Ftests%5C%2Fe2e-pw%5C%2Ftests%5C%2F%2F+test.describe.serial&type=code). However, the [official documentation discorages](https://playwright.dev/docs/api/class-test#test-describe-serial) the use of `test.describe.serial` in favor of `test.describe.configure`. I will create an issue to address the replacement of `test.describe.serial` in a follow-up PR.

> Can you add some more details and context on why we need this, and what caused the E2E tests to be flaky?

Good call, I'll update the PR description.

**Comment by rodelgc** [comment]
Thank you for updating these tests to conform with Playwright's preferred way of setting the `serial` mode.

However, I think the `serial` mode doesn't fully solve the conflict between the legacy product experience and the new one. As proof, if you go to the [E2E test report of this PR](https://woocommerce.github.io/woocommerce-test-reports/pr/39823/e2e/) and filter the tests that were retried, you would notice that the test `Enable block product editor > Enabled > is not hooked up to sidebar "Add New"` was retried. It failed on the first attempt, and passed on the second.

4diID9iv93

_(The other retried test, `Enable block product editor > Enabled > can enable the block product editor`, was retried only because it was previously skipped, which is a side-effect of the serial mode when an earlier test failed. So we can safely disregard the retry on this test.)_

When I looked at the cause of the failure of the first attempt, it's because the test was expecting to land on the legacy product editor after clicking `Products > Add New` from the navigation menu. But instead, it landed on the new product editor. Which means the conflict between the two product experiences weren't fully solved.

Here's how the Allure report reported the error:

jiXE2EYArJ

And here's the report from the Playwright trace file:

O7YIEM1EHe

I also encountered failures when I tested the PR locally, and for the same reason:

TwmRXE2SZQ

But when I removed `--repeat-each=8`, both sets of tests passed:

Miz303j4Yv

Serial mode only ensures the tests inside a test file were run sequentially, but it doesn't assure that test files won't overlap. 

@octaedro, would you consider separating the legacy and the new product experiences into separate Playwright projects? Currently, our E2E tests are all within a single project. My suggestion would be to have a second project that would only run the new product editor tests. And that this second project would depend on the first one (the one with the rest of the E2E tests ) using [Playwright project dependencies](https://playwright.dev/docs/test-projects#dependencies). This ensures the two experiences won't run in parallel in CI. Also, this approach could be a good precedent for upcoming experimental features.

**Review by louwie17 — APPROVED** [review]
Approving this as these changes don't hurt, but I believe there is still some flakiness happening.
My tests start failing around the ~50 mark which I guess is around the third run, pretty consistently, once that happens things start timing out pretty drastically.
So maybe there are some other improvements to be made?

Screenshot 2023-08-24 at 9 19 53 AM

**Comment by rodelgc** [comment]
I agree with @louwie17 that the changes in this PR still looks great and good to be merged, as it appropriately replaces the deprecated way of serializing tests with the more preferred one. Just noting that we will need a different approach to fully address the flakiness.

**Comment by louwie17** [comment]
> My suggestion would be to have a second project that would only run the new product editor tests. And that this second project would depend on the first one (the one with the rest of the E2E tests ) using [Playwright project dependencies](https://playwright.dev/docs/test-projects#dependencies). This ensures the two experiences won't run in parallel in CI. Also, this approach could be a good precedent for upcoming experimental features.

I like this suggestion a lot!

**Pull request merged**